### PR TITLE
[WIP]: Global Contracts

### PR DESCRIPTION
**Pull request by pugachAG**

### CODE REVIEW

**Comment by flmel** [comment]
Hi @pugachAG – thank you for starting this proposal. As the moderator, I labeled this PR as "Needs author revision" because we assume you are still working on it since you submitted it in "Draft" mode.
Please ping the [@near/nep-moderators](https://github.com/orgs/near/teams/nep-moderators) once you are ready for us to review it.

File: neps/nep-0591.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+With the advent of chain signatures, the smart contract wallet use case will become more ubiquitous.
+As a result, it is very desirable to be able to reuse already deployed contract without having to pay for the storage cost again.
+
+Additionally global contracts cover the underlying use case for [NEP-491](https://github.com/near/NEPs/pull/491): https://github.com/near/nearcore/pull/12818.
+
+## Specification
+
+Global contract can be deployed in 2 ways: either by its hash or by owner account id.
+Contracts deployed by hash are effectively immutable and cannot be updated.
+When deployed by account id the owner can redeploy the contract updating it for all its users.
```
Comment (walnut-the-cat) [inline neps/nep-0591.md:37]: Can we talk about when to use each way? For instance, when should a user use hash instead of account id, and vice versa?

File: neps/nep-0591.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+One potential issue is increasing infrastructure cost for global contracts with growing number of shards.
+A global contract is effectively replicated on every shard, so with increase in number of shards each global contract uses more storage.
+This can be potentially addressed in the future by making deployment costs parametrized with the number of shards in the current epoch, but it still wouldn't address the issue for the already deployed contracts.
+
+## Alternatives
+
+In [the original proposal](https://github.com/near/NEPs/issues/556) we considered storing global contracts in a separate global trie (managed at the block level) and introducing a dedicated distribution mechanism.
+We decided not to proceed with this approach as it requires significantly higher effort to implement and also introduces new critical dependencies for the protocol.
+
```
Comment (walnut-the-cat) [inline neps/nep-0591.md:162]: Did you intentionally skip Future possibilities and Consequences section? (e.g. https://github.com/near/NEPs/blob/global-contracts/neps/nep-0539.md#future-possibilities)

File: neps/nep-0591.md (hunk 1 above)
Comment (pugachAG) [reply neps/nep-0591.md:37]: Sure, added in bfad2e3

File: neps/nep-0591.md (hunk 2 above)
Comment (pugachAG) [reply neps/nep-0591.md:175]: Yes, I don't have anything to say in those section, so it was omitted

Comment (pugachAG) [reply neps/nep-0591.md:175]: I've added Future possibilities section in 053cb24 mentioning sharded contracts

**Comment by walnut-the-cat** [comment]
Thank you @pugachAG and @stedfn for submitting this NEP.

As a moderator, I reviewed this NEP and it meets [the proposed template guidelines](https://github.com/near/NEPs/blob/master/nep-0000-template.md). I am moving this NEP to [the REVIEW stage](https://github.com/near/NEPs/blob/master/neps/nep-0001.md#nep-lifecycle) and would like to ask the @near/wg-protocol working group members to assign 2 Technical Reviewers to complete a technical review (see expectations below).

Just for clarity, _Technical Reviewers play a crucial role in scaling NEAR ecosystem_ as they provide their in-depth expertise in the niche topic while work group members can stay on guard of the NEAR ecosystem. The discussions may get too deep and it would be inefficient for each WG member to dive into every single comment, so NEAR Developer Governance [designed this process](https://github.com/near/NEPs/blob/master/neps/nep-0001.md) that includes subject matter experts helping us to scale by writing a summary with the raised concerns and how they were addressed.

<details>

<summary>Technical Review Guidelines</summary>

 

* First, review the proposal within one week. If you have any suggestions that could be fixed, leave them as comments to the author. It may take a couple of iterations to resolve any open comments. 

* Second, once all the suggestions are addressed, produce a Technical Summary, which helps the working group members make a weighted decision faster. Without the summary, the working group will have to read the whole discussion and potentially miss some details.

Technical Summary guidelines:

  * A recommendation for the working group if the NEP is ready for voting (it could be approving or rejecting recommendation). Please note that this is the reviewer's personal recommendation.

  * A summary of benefits that surfaced in previous discussions. This should include a concise list of all the benefits that others raised, not just the ones that the reviewer personally agrees with.

  *  A summary of concerns or blockers, along with their current status and resolution. Again, this should reflect the collective view of all commenters, not just the reviewer's perspective.

Here is a [nice example](https://github.com/near/NEPs/pull/399#issuecomment-1462492128) and a template for your convenience:

```

### Recommendation

Add recommendation

### Benefits

* Benefit

* Benefit

### Concerns

| # | Concern | Resolution | Status |

| - | - | - | - |

| 1 | Concern | Resolution | Status |

| 2 | Concern | Resolution | Status |

```

Please tag the @near/nep-moderators once you are done, so we can move this NEP to the voting stage. Thanks again.

</details>

**Comment by bowenwang1996** [comment]
As a working group member, I nominate @akhi3030 and @wacban as SME reviewers

**Comment by akhi3030** [comment]
To help me schedule the work here, what is the deadline for this review?

File: neps/nep-0591.md (hunk 1 above)
Comment (akorchyn) [reply neps/nep-0591.md:37]: Sorry to jump in, but I’m curious, will the deployed-by account ID receive a hash?

I don’t see a reason to allow someone to change my code and potentially take everything from my contract. But if a globally deployed account also gets a hash that I can lock onto (and update later if needed), that would be a nice DevX improvement.

I do see a downside, though that every update essentially creates a new copy of the code instead of replacing it. So I’m curious, how is this going to be addressed?

**Comment by pugachAG** [comment]
@akhi3030 testnet release is scheduled for March 17th, so it would be nice to get this NEP reviewed by that that date.

File: neps/nep-0591.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+## Motivation
+
+A common use case on NEAR is to deploy the same smart contract many times on many different accounts. For example, a multisig contract is a frequently deployed contract.
+However, today each time such a contract is deployed, a user has to pay for its storage and the cost is quite high. For a 300kb contract the cost is 3N.
+
+With the advent of chain signatures, the smart contract wallet use case will become more ubiquitous.
+As a result, it is very desirable to be able to reuse already deployed contract without having to pay for the storage cost again.
+
+Additionally global contracts cover the underlying use case for [NEP-491](https://github.com/near/NEPs/pull/491): https://github.com/near/nearcore/pull/12818.
```
Comment (wacban) [inline neps/nep-0591.md:31]: Can you add a short context and explanation?

File: neps/nep-0591.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+struct GlobalContractData {
+    code: Vec<u8>,
+    id: GlobalContractIdentifier,
+}
+```
+
+`GlobalContractDistribution` receipt is generated as a result of processing `DeployGlobalContractAction`.
+Receipt distribution logic is updated to route such receipts to all shards.
+So effectively it is a part of `ShardProof` (incoming receipts) for each shard, but occurs only once in the outgoing receipts of that chunk.
+Applying `GlobalContractDistribution` receipt updates the corresponding `TrieKey::GlobalContractCode` in the trie.
```
Comment (wacban) [inline neps/nep-0591.md:117]: It would be good to test that this logic works well in case of resharding. One particular corner case would be:
* `#10` - deploy smart contract in shard X
* `#11` - shard Y and shard Z are both missing chunk
* `#12` - new shard layout after resharding where Y was split into Y1 and Y2

For X = Y, X = Z, X != Y && X != Z

File: neps/nep-0591.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+```
+
+Applying `UseGlobalContractAction` updates user account `contract` field accordingly.
+
+`FunctionCall` action processing is updated to respect global contracts. This includes updating [contract preparation pipeline](https://github.com/near/nearcore/blob/fb95d7b7740d1fda9245afa498ce4e9ac145c8af/runtime/runtime/src/pipelining.rs#L24) as well as [recording of the executed contract to be included in the state witness](https://github.com/near/nearcore/blob/fb95d7b7740d1fda9245afa498ce4e9ac145c8af/core/store/src/trie/update.rs#L338).
+
+### Costs
+
+For global contracts deployment we burn tokens for storage instead of locking like what we do regular contracts today.
+The cost per byte of global contract code `global_contract_storage_amount_per_byte` is set as 10x the storage staking cost `storage_amount_per_byte`.
```
Comment (wacban) [inline neps/nep-0591.md:140]: Is it set to 10x symbolically or do you just set its value to be 10x? I would suggest decoupling them so that they can be updated independently.

File: neps/nep-0591.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+Additionally we add action costs for the global contract related actions:
+
+* `action_deploy_global_contract` is exactly the same as `action_deploy_contract`
+* `action_deploy_global_contract_per_byte`:
+  * send costs are the same as `action_deploy_contract_per_byte`
+  * execution costs should cover distribution of the contract to all shards:
+    * this is pretty expensive for the network, so want want to change significant amount of gas for that
```
Comment (wacban) [inline neps/nep-0591.md:148]: typo: change -> charge

File: neps/nep-0591.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+enum GlobalContractDeployMode {
+    /// Contract is deployed under its code hash.
+    /// Users will be able reference it by that hash.
+    /// This effectively makes the contract immutable.
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
```
Comment (wacban) [inline neps/nep-0591.md:57]: Does deploying a global contract by account id deploy said contract to the owner account as well?

File: neps/nep-0591.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    AccountId(AccountId),
+}
+```
+
+The value is contract code bytes, similar to `TrieKey::ContractCode`.
+
+### Distribution
+
+Global contract has to be distributed to all shards after being deployed.
+This is implemented with a dedicated receipt type:
```
Comment (wacban) [inline neps/nep-0591.md:100]: Is this accounted for in the BandwidthScheduler?

**Review by wacban — COMMENTED** [review]
I had the first look, looks good so far, I'll come back to it some time this week.

File: neps/nep-0591.md (hunk 3 above)
Comment (pugachAG) [reply neps/nep-0591.md:31]: @staffik can you please help me here

File: neps/nep-0591.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
+}
+```
+
+Also new action is added for using previously deployed global contract:
+
```
[the remark this answers is not in the document]
Comment (pugachAG) [reply neps/nep-0591.md:62]: As far as I know it is not possible to unlink the regular contract, so we don't have to solve that for the global contracts. The same applies for the state of the contracts.

File: neps/nep-0591.md (hunk 5 above)
Comment (akhi3030) [reply neps/nep-0591.md:140]: +1 to this comment.

**Review by akhi3030 — APPROVED** [review]
generally looks good to me and I am happy for it to proceed with the current spec.  I will do a detailed SME review soon.

File: neps/nep-0591.md (hunk 3 above)
Comment (staffik) [reply neps/nep-0591.md:31]:
```suggestion
Additionally, global contracts cover the underlying use case for [NEP-491](https://github.com/near/NEPs/pull/491): [#12818](https://github.com/near/nearcore/pull/12818).
Some businesses onboard their users by creating an account on their behalf and deploying a contract to it. Such a use case is susceptible to refund abuse, where there is a financial incentive to repeatedly create and destroy an account, cashing in on the storage fee that was initially paid by the business to deploy the contract.
```
cc @pugachAG

File: neps/nep-0591.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    /// This effectively makes the contract immutable.
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
+}
+```
+
+Also new action is added for using previously deployed global contract:
```
[the remark this answers is not in the document]
Comment (pugachAG) [reply neps/nep-0591.md:64]: addressed in c528f22

File: neps/nep-0591.md (hunk 5 above)
Comment (pugachAG) [reply neps/nep-0591.md:145]: `global_contract_storage_amount_per_byte` is a separate runtime parameter completely independent from `storage_amount_per_byte`, here I meant that initial values is set to be `10 * storage_amount_per_byte`

File: neps/nep-0591.md (hunk 7 above)
Comment (pugachAG) [reply neps/nep-0591.md:58]: nope, it does not do that

File: neps/nep-0591.md (hunk 8 above)
Comment (pugachAG) [reply neps/nep-0591.md:105]: implemented as part of https://github.com/near/nearcore/pull/13147

File: neps/nep-0591.md (hunk 1 above)
Comment (pugachAG) [reply neps/nep-0591.md:38]: Nope, in this case we store account id, so the owner can update contract for all current users. There are underlying use cases for that, for example application developers onboarding users into crypto. 
Also it is possible to lock such contract if owner deletes all keys from the account, effectively making it impossible to make any changes.

File: neps/nep-0591.md [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    target_shard: ShardId,
+    already_delivered_shards: Vec<ShardId>,
+    code: Arc<[u8]>,
+}
+```
+
+`GlobalContractDistribution` receipt is generated as a result of processing `DeployGlobalContractAction`.
+Such receipt contains destination shard `target_shard` as well as list of already visited shard ids `already_delivered_shards`.
+Applying `GlobalContractDistribution` receipt updates the corresponding `TrieKey::GlobalContractCode` in the trie for the current shard.
+It also generates distribution receipt for the next shard in the current shard layout. This process continues until `already_delivered_shards` contains all shards.
```
Comment (wacban) [inline neps/nep-0591.md:128]: One consequence of this approach is that global contract deployment will take at least num_shards blocks to get onto all shards. In the future we may need some way of checking if the contract is fully deployed / updated. Nothing to worry about now, I'm sure the contract developers will ask about it once they need it.

File: neps/nep-0591.md [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+`AccountV2` is introduced changing `code_hash: CryptoHash` field to more generic `contract: AccountContract`:
+
+```rust
+enum AccountContract {
+    None,
+    Local(CryptoHash),
+    Global(CryptoHash),
+    GlobalByAccount(AccountId),
+}
+```
```
Comment (wacban) [inline neps/nep-0591.md:145]: Will it be possible to deploy / use global contracts from accounts created before AccountV2 is released? Is there any migration / lazy update necessary?

**Review by wacban — APPROVED** [review]
As a subject matter expert I approve this NEP. 

It adds an important feature, it is well designed, it strikes a great balance between added complexity and practical aspects and it well integrates with the rest of nearcore. Thank you @bowenwang1996, @pugachAG and @stedfn for this great proposal.

File: neps/nep-0591.md (hunk 11 above)
Comment (pugachAG) [reply neps/nep-0591.md:128]: Indeed with more shards in the future propagating global contract updates might take quite some time. 
We can extend the functionality in the future by adding some deployment progress indicator as part of owners account id data if that would be required.

File: neps/nep-0591.md [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    contract_identifier: GlobalContractIdentifier,
+}
+
+enum GlobalContractIdentifier {
+    CodeHash(CryptoHash),
+    AccountId(AccountId),
+}
+```
+
+This action is similar to deploying a regular contract to an account, except user does not cover storage deposit.
```
Comment (birchmd) [inline neps/nep-0591.md:77]: I think it is worth discussing possible errors arising from this action. If a code hash is given that does not correspond to a known global contract does the action fail? Similarly, if there is no global contract associated with the given account does the action fail?

This is important to know because global contracts take time to propagate to all shards and so a code hash which is unrecognized at one time might become recognized later.

File: neps/nep-0591.md [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+    /// This effectively makes the contract immutable.
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
+}
+```
+
+User pays for storage by burning NEAR tokens from its balance depending on the deployed contract size.
```
Comment (birchmd) [inline neps/nep-0591.md:62]: Error cases should be discussed here too. If the given code is not valid wasm are the Near tokens still burned? Or is a different amount of gas charged instead?

**Review by birchmd — APPROVED** [review]
As a working group member, I lean towards approving this proposal. In addition to the reduced cost of the "factory pattern", I especially look forward to how this feature lays the foundation for "sharded contracts" in the future.

I've left a few comments where I think this specification needs a little more detail.

**Review by akhi3030 — APPROVED** [review]
As a SME reviewer, I also lean towards approving this NEP.

Pros:
- it reduces the costs for deploying the same wasm module on multiple accounts making it cheaper for many people to use the network.
- it is a necessary requirement for sharded contracts

Cons:
- There are concerns about potential state not being cleaned up from the chain.  However, given the current storage staking incentives, we have similar problems already and in this instance the problem is less severe as the costs for deploying a global contract are high.

File: neps/nep-0591.md (hunk 14 above)
Comment (pugachAG) [reply neps/nep-0591.md:62]: good point, mentioned in 57545f524d3bf7eb9ed60c4db1cd64b28f5b6138

File: neps/nep-0591.md (hunk 13 above)
Comment (pugachAG) [reply neps/nep-0591.md:78]: added description in dcfdcd3bfffe253234cf1bf68e760f27330df50f, thanks!

**Comment by pugachAG** [comment]
@birchmd thanks for your review!

> It might be worth adding a note on how this routing interacts with the new instant-resharding feature.

I've added description for that in 6bf2c85c893dbac416eb8c99201541834e02310a

File: neps/nep-0591.md (hunk 11 above)
Comment (pugachAG) [reply neps/nep-0591.md:130]: added resharding description in 6bf2c85c893dbac416eb8c99201541834e02310a

**Comment by flmel** [comment]
@near/wg-protocol Can you please comment in the thread if you are leaning towards approving or rejecting this NEP? Please include your rationale and/or any feedback that you have for the authors.

File: neps/nep-0591.md [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+    already_delivered_shards: Vec<ShardId>,
+    code: Arc<[u8]>,
+}
+```
+
+`GlobalContractDistribution` receipt is generated as a result of processing `DeployGlobalContractAction`.
+Such receipt contains destination shard `target_shard` as well as list of already visited shard ids `already_delivered_shards`.
+Applying `GlobalContractDistribution` receipt updates the corresponding `TrieKey::GlobalContractCode` in the trie for the current shard.
+It also generates distribution receipt for the next shard in the current shard layout. This process continues until `already_delivered_shards` contains all shards.
+This way we ensure that at any point in time there is at most one `GlobalContractDistribution` receipt in flight for a given deployment and eventually it will reach all shards.
```
Comment (mfornet) [inline neps/nep-0591.md:131]: > This way we ensure that at any point in time there is at most one `GlobalContractDistribution` receipt in flight

What is the motivation for having at most one such receipt, instead of deploying it in parallel to all shards, or at least using some other strategy that allows faster deployment.

File: neps/nep-0591.md [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+}
+```
+
+Applying `UseGlobalContractAction` updates user account `contract` field accordingly.
+
+`FunctionCall` action processing is updated to respect global contracts. This includes updating [contract preparation pipeline](https://github.com/near/nearcore/blob/fb95d7b7740d1fda9245afa498ce4e9ac145c8af/runtime/runtime/src/pipelining.rs#L24) as well as [recording of the executed contract to be included in the state witness](https://github.com/near/nearcore/blob/fb95d7b7740d1fda9245afa498ce4e9ac145c8af/core/store/src/trie/update.rs#L338).
+
+### Costs
+
+For global contracts deployment we burn tokens for storage instead of locking like what we do regular contracts today.
```
Comment (mfornet) [inline neps/nep-0591.md:156]: Do we burn the tokens for storage on every deployment? even when we are using deploying by account id which overrides and removes the previous one?

File: neps/nep-0591.md [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+    id: GlobalContractIdentifier,
+    target_shard: ShardId,
+    already_delivered_shards: Vec<ShardId>,
+    code: Arc<[u8]>,
+}
+```
+
+`GlobalContractDistribution` receipt is generated as a result of processing `DeployGlobalContractAction`.
+Such receipt contains destination shard `target_shard` as well as list of already visited shard ids `already_delivered_shards`.
+Applying `GlobalContractDistribution` receipt updates the corresponding `TrieKey::GlobalContractCode` in the trie for the current shard.
```
Comment (mfornet) [inline neps/nep-0591.md:129]: Just to clarify, this means that a contract becomes available in a shard after processing the `DeployGlobalContractAction` action, so the contract can be used in some shards before other ones right?

Comment (mfornet) [reply neps/nep-0591.md:129]: In case this is correct, some "funny" side effect, is that two contracts living in different shards, pointing at the same global contract, could run an action on the same block, and be using two different contracts, since the update is still in flight.

By "funny" I mean that it is hard to foresee side effects of this behavior, or if it could be exploited.

**Review by mfornet — APPROVED** [review]
Congrats on delivering this NEP. 

As a working group member I lean toward approving this NEP, and personally I'm looking forward to being released.

I left a few comments and questions.

File: neps/nep-0591.md (hunk 15 above)
Comment (pugachAG) [reply neps/nep-0591.md:131]: Having at most one receipt allows us to easily handle edge case when deployment happens at the resharding boundary and we need to make sure the contract code reaches shards introduced in the new layout.
When deploying to all shards at once (in parallel) resharding would require either special handling or major changes to how we route buffered receipts. Specifically when buffered deployment receipt is targeted to an old shard, we can only re-route it to only one of the child shards after resharding, hence special logic would be required when applying that receipt to further forward it to the other child shard.
To summarise: we considered multiple options and decided to proceed with the current approach as it was the easiest to implement and reason about

File: neps/nep-0591.md (hunk 16 above)
Comment (pugachAG) [reply neps/nep-0591.md:156]: Yes, in the initial implementation we decided to proceed with the simplest approach and always burn tokens regardless of previously deployed global contract under that account id. This also indirectly covers distribution costs that cannot be fully accounted in gas costs due to chunk gas limit.

File: neps/nep-0591.md (hunk 17 above)
Comment (pugachAG) [reply neps/nep-0591.md:129]: You are right, this can result in such "funny" behaviours. Effectively during upgrade 2 versions of the contract will be active on different shards for some range of blocks. That is something contract developers should be aware of. Bowen also raised that point and we reached a conclusion that it should not be a big issue in practice.

File: neps/nep-0591.md (hunk 16 above)
Comment (bowenwang1996) [reply neps/nep-0591.md:156]: The previous one isn't really removed. It will still be there

Comment (pugachAG) [reply neps/nep-0591.md:156]: In case of deployment by account_id the previously deployed code will be replaced by the new one

**Comment by bowenwang1996** [comment]
As a working group member, I approve this NEP

**Comment by pugachAG** [comment]
@flmel I guess this can be merged now?

**Review by flmel — APPROVED** [review]
